
<img src="http://oci02.img.iteso.mx/identidad_de_instancia_2018/ITESO/Logos%20ITESO/Logo-ITESO-Principal.jpg">

<center> <font color= #004A94> <font size = 6> Lic. Ingeniería Financiera </font> <br> <br> <font color= #047CFB> <font size = 4>I.F. Francisco Ricardo Rodriguez Ramirez - if705783@iteso.mx </font>

<center><font color= #004A94> <font size = 6> Microestructura y Sistemas de Trading </font> <br> <br>

<center><font color=#42c29b><font size=5> Laboratorio 2: Behavioral Finance </font> <br> <br>

###### <span style="font-family:Papyrus; font-size:4em;"> Resumen 


#### Objetivo: Analizar el comportamiento financiero de un Trader 
    

#### Especificaciones: 
* Se analiza el archivo obtenido de MetraTrader4 el cual contiene los datos histróricos durante el lapso que inicia el 11/02/2020 y finaliza el 27/02/2020 
   


In [11]:
from bokeh.io import output_notebook, show
output_notebook()
from pandas import __version__ as pandas_version
from plotly import __version__ as ploty

Loading BokehJS ...

###### <span style="font-family:Papyrus; font-size:4em;"> Código

A continuación se presentan las librerias que fueron utilizadas
* Pandas : Utilizada para facilitar la visualización de datos, cumple con el objetivo de facilitar el manejo de estos
* Numpy  : Utilizada para realizar cálculos, en este caso utilizada para sacar rendimientos logartimicos y desviación estándar.
    

Además los archivos Python File desarrollados para cumplir con el objetivo de la clase.
* Funciones: Desarrollo de funciones para el respectivo análisis que será descrito a más adelante.
* Principal: Visualización de las variables con respecto a las funciones desarrolladas en el archivo anterior.7
* Visualizaciones: Cumple con el objetivo de hacer llamar las funciones para las gráficas.




In [12]:
%%capture 
import datos
import funciones as fn
import visualizaciones as vn
import numpy as np
import principal as pr
import plotly.express as px

###### <span style="font-family:Papyrus; font-size:4em;"> Visualización de datos

Del Python File principal se manda llamar la variable df_data el cual corresponde a el DataFrame de los datos históricos del Trader. En este se puede visualizar desde el número y el tipo de orden es decir si fue de venta o de compra, la fechas tanto de cuando se inició la operación como de cuanto se cerró, las ganancias o pérdidas y además se añadieron columnas como el capital acumulado recordadno que en este caso fue de 5,000 dólares </p>

In [13]:
df_data=pr.df_data
df_data

order            opentime  type  size     symbol   openprice  \
0   233205643 2020-02-11 18:37:26   buy  0.01     eurusd     1.09205   
1   233212393 2020-02-11 21:08:03   buy  0.01     eurusd     1.09226   
2   233359508 2020-02-13 03:12:17  sell  0.01     eurusd     1.08741   
3   233403856 2020-02-13 15:23:26   buy  0.01     eurusd     1.08661   
4   233403897 2020-02-13 15:24:22  sell  0.01     eurusd     1.08628   
5   233565329 2020-02-17 18:37:51   buy  0.01     bcousd    57.65300   
6   233736505 2020-02-19 18:52:46   buy  0.10     bcousd    59.37200   
7   233737244 2020-02-19 19:08:44   buy  0.10   wticousd    53.19900   
8   233737426 2020-02-19 19:13:13   buy  0.10    mbtcusd   102.44300   
9   233737673 2020-02-19 19:19:16   buy  0.03    cornusd     3.77900   
10  233739710 2020-02-19 20:24:26   buy  0.05  spx500usd  3391.20000   
11  233739987 2020-02-19 20:43:35   buy  0.50  spx500usd  3390.70000   
12  233740424 2020-02-19 20:59:50   buy  1.00  spx500usd  3390.70000   
13  233815084 2020-02-20 15:35:26  sell  0.02     eurusd     1.07925   
14  233827005 2020-02-20 18:49:49   buy  0.01     xauusd  1622.84900   
15  233829098 2020-02-20 19:22:20   buy  1.00  spx500usd  3352.90000   
16  233829139 2020-02-20 19:23:18   buy  1.00  spx500usd  3352.40000   
17  233829246 2020-02-20 19:29:23   buy  1.00  spx500usd  3352.20000   
18  233829362 2020-02-20 19:32:21   buy  1.50  spx500usd  3351.20000   
19  233829482 2020-02-20 19:37:20   buy  1.50  spx500usd  3354.80000   
20  233829552 2020-02-20 19:41:25   buy  1.50  spx500usd  3354.90000   
21  233830296 2020-02-20 20:03:54   buy  1.50  spx500usd  3358.90000   
22  233830743 2020-02-20 20:18:20   buy  1.00  spx500usd  3366.40000   
23  233830778 2020-02-20 20:19:52   buy  0.50  spx500usd  3364.40000   
24  233830995 2020-02-20 20:31:13  sell  1.00  spx500usd  3365.60000   
25  233831333 2020-02-20 20:38:34   buy  1.50  spx500usd  3362.90000   
26  233831369 2020-02-20 20:41:22  sell  1.00  spx500usd  3362.60000   
27  233831457 2020-02-20 20:45:35   buy  1.50  spx500usd  3362.20000   
28  233831482 2020-02-20 20:47:38   buy  1.25  spx500usd  3364.40000   
29  233831503 2020-02-20 20:49:02  sell  1.50  spx500usd  3363.20000   
30  233831645 2020-02-20 20:57:45  sell  1.00  spx500usd  3361.40000   
31  233831991 2020-02-20 21:07:45  sell  0.50  spx500usd  3367.60000   
32  233832221 2020-02-20 21:20:20   buy  0.50  spx500usd  3368.70000   
33  233832466 2020-02-20 21:30:11   buy  1.00  spx500usd  3368.50000   
34  233832609 2020-02-20 21:36:39   buy  1.00  spx500usd  3374.20000   
35  233832663 2020-02-20 21:39:31   buy  1.50  spx500usd  3373.70000   
36  233832927 2020-02-20 21:54:43   buy  1.50  spx500usd  3369.90000   
37  233832986 2020-02-20 21:58:55   buy  1.00  spx500usd  3371.40000   
38  233833672 2020-02-20 22:37:33   buy  1.00  spx500usd  3372.70000   

           s/l         t/p           closetime  closeprice  ...  taxes  swap  \
0      1.09182     1.09245 2020-02-11 18:38:30     1.09182  ...      0     0   
1      0.00000     0.00000 2020-02-13 03:14:53     1.08738  ...      0     0   
2      0.00000     0.00000 2020-02-13 03:29:56     1.08731  ...      0     0   
3      0.00000     0.00000 2020-02-13 15:28:09     1.08601  ...      0     0   
4      0.00000     0.00000 2020-02-13 16:40:33     1.08621  ...      0     0   
5     53.77000    59.50000 2020-02-19 18:29:09    59.50100  ...      0     0   
6     53.70000    64.35000 2020-02-20 16:25:18    59.73500  ...      0     0   
7     49.52000    57.35000 2020-02-20 16:16:54    53.82300  ...      0     0   
8    101.94300   103.44300 2020-02-19 19:32:10   101.93800  ...      0     0   
9      3.64200     4.00000 2020-02-27 16:30:25     3.64200  ...      0     0   
10  3390.20000  3441.20000 2020-02-19 20:38:18  3390.10000  ...      0     0   
11  3389.70000     0.00000 2020-02-19 20:58:39  3389.50000  ...      0     0   
12     0.00000     0.00000 2020-02-19 21:14:44  3391.60000  ...      0     0   
13     

###### <span style="font-family:Papyrus; font-size:4em;"> Estadística Básica

<p style="text-align: justify;"> Del Python File principal se manda llamar la variable df_1_tabla el cual pertenece al DataFrame con la estadística básica de las operaciones realizadas. Será descrito a continuación.</p>

In [14]:
pr.df_1_tabla

medida  valor                             descripcion
0       Ops totales  39.00                     Operaciones totales
1         Ganadoras  27.00                   Operaciones ganadoras
2       Ganadoras_c  20.00         Operaciones ganadoras de compra
3       Ganadoras_v   7.00          Operaciones ganadoras de venta
4        Perdedoras  12.00                  Operaciones perdedoras
5      Perdedoras_c  11.00        Operaciones perdedoras de compra
6      Perdedoras_v   1.00         Operaciones perdedoras de venta
7     Media(Profit)  50.00           Mediana profit de operaciones
8       Media(Pips)  -5.00          Mediana de pips de operaciones
9     r_efectividad   0.69    Ganadoras Totaes/Operaciones Totales
10     r_proporcion   2.25  Perdedoras Totales/Operaciones Totales
11  r_efectividad_c   0.51   Ganadoras Compras/Operaciones Totales
12  r_efectividad_v   0.18    Ganadoras Ventas/Operaciones Totales

De la tabla anterior podemos determinar lo siguiente:
* De un total de 39 operaciones ralizadas: 
    * Del 69% de las operaciones terminaron como ganancias. 
    * El 31% de las operaciones terminaron como pérdidas. 
    * Las operaciones de compra tuvieron mayor efectividad sobre las de venta.
    * Se obtuvo un promedio de 50 dólares por operación como ganancia. 


###### <span style="font-family:Papyrus; font-size:4em;"> Ranking Ratio de Efectividad

<p style="text-align: justify;">Del Python File principal se manda llamar la variable df_1_ranking el cual pertenece al DataFrame con el ratio de efectivada de los instrumentos utilizados por el trader</p>

In [15]:
df = pr.df_1_ranking
df

simbolo    rank
0     bcousd  100.00
1   wticousd  100.00
2  spx500usd   77.78
3     eurusd   50.00
4     xauusd    0.00
5    cornusd    0.00
6    mbtcusd    0.00

In [16]:
figure = pr.grafica_pastel
figure

<p style="text-align: justify;"> Los instrumentos que tuvieron mayor efectividad fueron el BCOUSD el cual representa el BridgeCoin  y el WTICOUSD que hace referencia al crudo de petróleo. Sin embargo si analizamos en el histórico de datos podemos observar que se realizaron minimas operaciones, mismo caso para los que tuvieron una efectividad de 0 en donde únicamente se realizó una operación.            Este trader utilizó principalmente los instrumentos del índice SPX500 y el tipo de cambio EURODOLAR, de los cuales obtuvo un ratio de efectividad de 77.78% y de 50% respectivamente </p>

###### <span style="font-family:Papyrus; font-size:4em;"> Profit diario

<p style="text-align: justify;">Del Python File principal se manda llamar la variable df_profit_diario el cual pertenece al DataFrame que contiene en la columna profit_d las ganancias díarias y en la columna profit_acm_d el acumulado de las ganancias. Recordando que el trader inició con un capital de $5,000 dólares.
En este DataFrame se incluye todos los días del calendario, con respecto a los días transcurridos desde la primera hasta la última operación e inclusive se toman en cuenta los sábados.</p>

In [17]:
pr.df_profit_diario

timestamp  profit_d  profit_acm_d
0  2020-02-11     -0.23       4999.77
1  2020-02-12      0.00       4999.77
2  2020-02-13     -5.31       4994.46
3  2020-02-14      0.00       4994.46
4  2020-02-15      0.00       4994.46
5  2020-02-16      0.00       4994.46
6  2020-02-17      0.00       4994.46
7  2020-02-18      0.00       4994.46
8  2020-02-19     21.30       5015.76
9  2020-02-20   1031.17       6046.93
10 2020-02-21      0.00       6046.93
11 2020-02-22      0.00       6046.93
12 2020-02-23      0.00       6046.93
13 2020-02-24      0.00       6046.93
14 2020-02-25      0.00       6046.93
15 2020-02-26      0.00       6046.93
16 2020-02-27    -20.55       6026.38

<p style="text-align: justify;"> Podemos observar que la ganancia mayor por día fue de 1,031.17 dólares en donde se tuvo un acumulado de 6043.93 dólares. A continuación se analizarán los movimientos que se realizaron ese día.</p>

In [18]:
filtro0 =df_data[df_data['dias_close']==20]
filtro0


order            opentime  type  size     symbol   openprice  \
6   233736505 2020-02-19 18:52:46   buy  0.10     bcousd    59.37200   
7   233737244 2020-02-19 19:08:44   buy  0.10   wticousd    53.19900   
13  233815084 2020-02-20 15:35:26  sell  0.02     eurusd     1.07925   
14  233827005 2020-02-20 18:49:49   buy  0.01     xauusd  1622.84900   
15  233829098 2020-02-20 19:22:20   buy  1.00  spx500usd  3352.90000   
16  233829139 2020-02-20 19:23:18   buy  1.00  spx500usd  3352.40000   
17  233829246 2020-02-20 19:29:23   buy  1.00  spx500usd  3352.20000   
18  233829362 2020-02-20 19:32:21   buy  1.50  spx500usd  3351.20000   
19  233829482 2020-02-20 19:37:20   buy  1.50  spx500usd  3354.80000   
20  233829552 2020-02-20 19:41:25   buy  1.50  spx500usd  3354.90000   
21  233830296 2020-02-20 20:03:54   buy  1.50  spx500usd  3358.90000   
22  233830743 2020-02-20 20:18:20   buy  1.00  spx500usd  3366.40000   
23  233830778 2020-02-20 20:19:52   buy  0.50  spx500usd  3364.40000   
24  233830995 2020-02-20 20:31:13  sell  1.00  spx500usd  3365.60000   
25  233831333 2020-02-20 20:38:34   buy  1.50  spx500usd  3362.90000   
26  233831369 2020-02-20 20:41:22  sell  1.00  spx500usd  3362.60000   
27  233831457 2020-02-20 20:45:35   buy  1.50  spx500usd  3362.20000   
28  233831482 2020-02-20 20:47:38   buy  1.25  spx500usd  3364.40000   
29  233831503 2020-02-20 20:49:02  sell  1.50  spx500usd  3363.20000   
30  233831645 2020-02-20 20:57:45  sell  1.00  spx500usd  3361.40000   
31  233831991 2020-02-20 21:07:45  sell  0.50  spx500usd  3367.60000   
32  233832221 2020-02-20 21:20:20   buy  0.50  spx500usd  3368.70000   
33  233832466 2020-02-20 21:30:11   buy  1.00  spx500usd  3368.50000   
34  233832609 2020-02-20 21:36:39   buy  1.00  spx500usd  3374.20000   
35  233832663 2020-02-20 21:39:31   buy  1.50  spx500usd  3373.70000   
36  233832927 2020-02-20 21:54:43   buy  1.50  spx500usd  3369.90000   
37  233832986 2020-02-20 21:58:55   buy  1.00  spx500usd  3371.40000   
38  233833672 2020-02-20 22:37:33   buy  1.00  spx500usd  3372.70000   

           s/l         t/p           closetime  closeprice  ...  taxes  swap  \
6     53.70000    64.35000 2020-02-20 16:25:18    59.73500  ...      0     0   
7     49.52000    57.35000 2020-02-20 16:16:54    53.82300  ...      0     0   
13     1.08549     1.06417 2020-02-20 22:07:07     1.07896  ...      0     0   
14  1620.84900     0.00000 2020-02-20 18:54:46  1620.83300  ...      0     0   
15  3352.40000     0.00000 2020-02-20 19:22:23  3352.10000  ...      0     0   
16     0.00000     0.00000 2020-02-20 19:28:00  3353.90000  ...      0     0   
17     0.00000     0.00000 2020-02-20 19:30:37  3354.90000  ...      0     0   
18     0.00000     0.00000 2020-02-20 19:33:59  3352.90000  ...      0     0   
19     0.00000     0.00000 2020-02-20 19:39:18  3355.70000  ...      0     0   
20     0.00000     0.00000 2020-02-20 19:47:54  3356.40000  ...      0     0   
21     0.00000     0.00000 2020-02-20 20:05:58  3359.90000  ...      0     0   
22     0.00000     0.00000 2020-02-20 20:34:28  3366.90000  ...      0     0   
23     0.00000     0.00000 2020-02-20 20:28:14  3366.60000  ...      0     0   
24     0.00000     0.00000 2020-02-20 20:33:37  3365.20000  ...      0     0   
25     0.00000     0.00000 2020-02-20 20:40:11  3363.40000  ...      0     0   
26     0.00000     0.00000 2020-02-20 20:42:07  3361.20000  ...      0     0   
27     0.00000     0.00000 2020-02-20 20:46:35  3362.90000  ...      0     0   
28     0.00000     0.00000 2020-02-20 21:01:24  3364.90000  ...      0     0   
29     0.00000     0.00000 2020-02-20 20:56:47  3362.40000  ...      0     0   
30     0.00000     0.00000 2020-02-20 22:48:42  3370.40000  ...      0     0   
31     0.00000     0.00000 2020-02-20 22:11:47  3365.40000  ...      0     0   
32     0.00000     0.00000 2020-02-20 21:33:07  3370.10000  ...      0     0   
33     0.00000     0.00000 2020-02-20 21:33:04  3370.40000  ...      0     0   

<p style="text-align: justify;"> Ese día se realizaron 28 operaciones, el instrumento que más predomino fue el ínidice del SPX500. Ese día el trader logró hacer 2,543.18 dólares de 23 operaciones. Sin embargo, de las 5 posiciones restante tuvo une pérdida de 1,512.01 dólares. Por lo que al final del día logro un acumulado de 1,031.17 dólares</p>

##  Métricas de desempeño

Del Python File principal se manda llamar la variable f_estadistica_basica el cual pertenece al DataFrame que contiene las siguientes métricas de desempeño.
* Sharpe 
* Sortino 
* Drawdown
* Drawup 
* Information ratio
  

In [19]:
estadistica = pr.f_estadistica_mad
estadistica

metrica                                valor  \
0         Sharpe                             0.252859   
1      Sortino_c                            Pendiente   
2      Sortino_v                            Pendiente   
3  Drawdown_capi     2020-02-11 | 2020-02-18 | $ 5.31   
4    Drawup_capi  2020-02-18 | 2020-02-26 | $ 1052.47   
5  Information_r                   0.4664965269390065   

                      descripcion  
0                    Sharpe Ratio  
1  Sortino para posiciones compra  
2   Sortino para posiciones venta  
3             DrawDown de capital  
4               DrawUp de capital  
5               Information_ratio